content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -579,41 +657,919 @@ sending values over the wire, the code should query the size of the
+focus on setting the size of objects. 
+
+We've said that the size of an object is not necessarily the same as the 
+size of the object's type. The object size won't be smaller, but it 
+could be larger. Why? For a stand-alone object or a parameter, most 
+implementations will round the size up to a storage element boundary, or 
+more, so the object size might be greater than that of the type. Think 
+back to Boolean, where Size is required to be 1, but stand-alone objects 
+are probably allocated 8 bits, i.e., an entire storage element (on our 
+hypothetical byte-addressed machine). 
+
+Likewise, recall that numeric type declarations are mapped to underlying 
+hardware numeric types. These underlying numeric types provide at least 
+the capabilities we request with our type declarations, e.g., the range 
+or number of digits, perhaps more. But the mapped numeric hardware type 
+cannot provide less than requested. If there is no underlying hardware 
+type with at least our requested capabilities, our declarations won't compile. 
+That mapping means that specifying the size of a numeric type doesn't  
+necessarily affect the size of objects of the type. That numeric 
+hardware type is the size that it is, and is fixed by the hardware. 
+
+For example, let's say we have this declaration:
+
+.. code-block:: ada
+
+   type Device_Register is range 0 .. 2**5 - 1 with Size => 5;
+   
+That will compile successfully, because there will be a signed integer 
+hardware type with at least that range. (Not necessarily, legally 
+speaking, but realistically speaking, there will be such a hardware 
+type.) Indeed, it may be an 8-bit signed integer, in which case 
+Device_Register'Size will give us 5, but objects of the type will have a 
+size of 8, unavoidably, even though we set Size to 5.
+
+The difference between the type and object sizes can lead to potentially 
+problematic code: 
+
+.. code-block:: ada
+
+   type Device_Register is range 0 .. 2**8 - 1 with Size => 8;
+   My_Device : Device_Register
+     with Address => To_Address (...); 
+
+The code compiles successfully, and tries to map a byte to a hardware 
+device that is physically connected to one storage element in the 
+processor memory space. The actual address is elided as it is not 
+important here. 
+
+That code might work too, but it might not. We might think that 
+My_Device'Size is 8, and that My_Device'Address points at an 8-bit 
+location. However, this isn't necessarily so, as we saw with the 
+supposedly 5-bit example earlier. Maybe the smallest signed integer the 
+hardware has is 16-bits wide. The code would compile because a 16-bit 
+signed numeric type can certainly handle the 8-bit range requested. 
+My_Device'Size would be then 16, and because 'Address gives us the 
+*starting* storage element, My_Device'Address might designate the 
+high-order byte of the overall 16-bit object. When the compiler reads 
+the two bytes for My_Device what will happen? One of the bytes will be 
```

Comment:
```suggestion
:ada:`My_Device'Size` would be then 16, and because :ada:`'Address` gives us the 
*starting* storage element, :ada:`My_Device'Address` might designate the 
high-order byte of the overall 16-bit object. When the compiler reads 
the two bytes for :ada:`My_Device` what will happen? One of the bytes will be 
```